### ISSUE TITLE
getReactNativePersistence breaks React Native Fast Refresh

### ISSUE THREAD

**Issue by Slapbox** [firebase/firebase-js-sdk#6231]
### Describe your environment

  * Operating System version: Windows 7
  * Browser version: N/A
  * Firebase SDK version: 9.6.7
  * Firebase Product: Auth

###  Describe the problem

Firebase@9.x Auth requires `getReactNativePersistence` on React Native - but including the relevant code for `getReactNativePersistence` causes the auth listener logic in our app to fire every time a change is made, crashing our app (without apparent error message) and preventing us from using Fast Refresh at all.

#### Steps to reproduce:

Unsure of full steps that may be required. For us, including the code below triggers it in logic that otherwise worked in Firebase@8.x [and in Firebase@9.x for a while.](https://github.com/firebase/firebase-js-sdk/issues/6050#issuecomment-1112799697)

#### Relevant Code:

```javascript
  initializeAuth(firebaseApp, {
    persistence: getReactNativePersistence(AsyncStorage),
  });
```

**Comment by Slapbox** [comment]
It may be that this only occurs during debugging but I can't confirm that at the moment now due to a major overhaul in our tooling.

**Comment by ishmaelmoreno** [comment]
@Slapbox does this still happen for you?

**Comment by Slapbox** [comment]
I think no, but I'm not sure of that. 

I don't think the possible resolution is related to any change in Firebase though, but rather that we abandoned react-native-debugger in favor of Flipper.

**Comment by Slapbox** [comment]
Yes, it does still happen, even with Flipper.

`Firebase: Error (auth/already-initialized).`

**Comment by sam-gc** [comment]
The problem is that `initializeAuth` and `getAuth` can be called multiple times but only if the underlying dependencies don't change. For example, calling `initializeAuth({persistence: browserLocalPersistence});` 10 times won't throw `auth/already-initialized`, but calling `initializeAuth({persistence: browserLocalPersistence}); initializeAuth({persistence: browserSessionPersistence});` _will_ throw the error since the persistence dependency has changed.

In this particular case, `getReactNativePersistence` returns a new object every time its called, so it appears `initializeAuth` is being called with different dependencies. It may be possible to avoid this behavior, but I can't give any promises as to timeline. In the meantime, you may be able to avoid this problem by disabling fast refresh specifically in the file that calls `initializeAuth` (e.g. https://github.com/vercel/next.js/issues/13268#issuecomment-743911710)

**Comment by Slapbox** [comment]
Thanks for your reply @sam-gc.

Our file that calls `initializeAuth` has no React exports, so I think that the workaround you linked probably cannot work for us.

This is the full file that calls `initializeAuth` in our app looks like below, and the dependencies for `initializeAuth` shouldn't change - unless the `firebaseApp` argument is changed? `firebaseApp` is the output of `initializeApp(config)`.

When the app is reloaded, `initializeApp(config)` presumably gets called anew and passed to `initializeAuth`:

```js
import { initializeAuth } from 'firebase/auth';
import { getReactNativePersistence } from 'firebase/auth/react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';

export const afterFirebaseInitialize = firebaseApp => {
  initializeAuth(firebaseApp, {
    persistence: getReactNativePersistence(AsyncStorage),
  });
};
```

**Comment by sam-gc** [comment]
When this gets reloaded, `getReactNativePersistence(AsyncStorage)` will return a different object -- that's where the issue lies. `firebaseApp` shouldn't be changing, since `initializeApp` similarly returns the same object for the same dependencies

**Comment by Slapbox** [comment]
It _shouldn't_ get reloaded as a new object in the code we have, but it could be. It's hard to know because the issue doesn't occur 100% of the time now with Flipper as our debugging tool and the quirks of React Native debugging.

Our code looks like this:

```js
let firebaseApp;

const makeFirebase = config => {
  if (firebaseApp) return firebaseApp;

  firebaseApp =
    typeof global.window !== 'undefined' ? initializeApp(config) : null;

  if (typeof afterFirebaseInitialize === 'function')
    afterFirebaseInitialize(firebaseApp);

  return firebaseApp;
};
```

**Comment by Slapbox** [comment]
@sam-gc why would the line `if (firebaseApp) return firebaseApp;` not prevent double initialization? I assume there's no code that would cause the results of `initializeApp(config)` to be cleared, and that's what's stored in `firebaseApp`. 

Something doesn't add up unless there's more to the story here.

**Comment by Slapbox** [comment]
This remains an issue and we're definitely not initializing anything twice. It's painstaking to be without Fast Refresh, and it's apparent we're not the only ones with the issue. It only began with `getReactNativePersistence` and there is exactly one place that `initializeAuth` is being called, and that's where we're defining the persistence method.

This is it. This is the sole place it's called.

```js
export const afterFirebaseInitialize = firebaseApp => {
  initializeAuth(firebaseApp, {
    persistence: getReactNativePersistence(AsyncStorage),
  });
};
```

I'd appreciate it so much if this was given some attention.

**Comment by ishmaelmoreno** [comment]
I've avoided updating my version of firebase due to this issue. I hope it is resolved soon because it defeats the purpose of using react native.

**Comment by Slapbox** [comment]
Friendly bump @sam-gc. 

I hope this will not go neglected. It's hard for me to fathom fixing a newly introduced bug not being "on the roadmap." An undocumented regression without a workaround is a bug that needs fixing in my book.

**Comment by Slapbox** [comment]
If it's true that the explanation for this is users double invoking `initializeAuth` or `getReactNativePersistence` then this code should work, but it doesn't:

```js
import moize from 'moize';
import { initializeApp } from 'firebase/app';
import { initializeAuth } from 'firebase/auth';
import { getReactNativePersistence } from 'firebase/auth/react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';

const persistence = {
  persistence: getReactNativePersistence(AsyncStorage),
};

const CONSTANT = 'CONSTANT';

const createMemoization = moize({
  maxSize: 1,
  maxArgs: 1, // Only memoize based on the first argument - ie never update because we're passing CONSTANT for first arg
});

const afterFirebaseInitialize = createMemoization((_unused, firebaseApp) => {
  initializeAuth(firebaseApp, persistence);
});

let firebaseApp;

const makeFirebase = config => {
  if (firebaseApp) return firebaseApp;

  firebaseApp = initializeApp(config);
  afterFirebaseInitialize(CONSTANT, firebaseApp);
};
```

The explanation that this is a React Fast Refresh problem would hold water if this didn't _only_ affect Firebase on React Native. No other projects on any other platform break like this, **and crucially, nor did Firebase used to break like this with v8 and below.**

**Comment by Slapbox** [comment]
If it's true that the explanation for this is users double invoking `initializeAuth` or `getReactNativePersistence` then this code should work, but it doesn't:

```javascript
import { initializeApp } from 'firebase/app';
import { initializeAuth } from 'firebase/auth';
import { getReactNativePersistence } from 'firebase/auth/react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';

let initialized = false;

const persistence = {
persistence: getReactNativePersistence(AsyncStorage),
};

export const afterFirebaseInitialize = firebaseApp => {
if (initialized) return;
initialized = true;
initializeAuth(firebaseApp, persistence);
};

let firebaseApp;

const makeFirebase = config => {
if (firebaseApp) return firebaseApp;

firebaseApp = initializeApp(config);
afterFirebaseInitialize(firebaseApp);
};
```

The explanation that this is a React Fast Refresh problem would hold water if this didn't _only_ affect Firebase on React Native. No other projects on any other platform break like this, **and crucially, nor did Firebase used to break like this with version 8 and below.**

[Even more crucially, Firebase 9 used to work as intended for us](https://github.com/firebase/firebase-js-sdk/issues/6050#issuecomment-1112799697). I'll have to go back and try to figure out more about why that may be - too late right now.

**Comment by hsubox76** [comment]
Oh! I just notiiced you're importing from both `firebase/auth` and `firebase/auth/react-native`. I think the second is actually a full copy of the first with a few modifications for RN, so I think you want to do just:

```
import { initializeAuth, getReactNativePersistence } from 'firebase/auth/react-native';
```

and get rid of the `firebase/auth` line. I'm not sure if this will fix your issue but can you give a try and see if it does? If it does, of course it seems we have some docs work to address.

**Comment by ishmaelmoreno** [comment]
Hi @Slapbox did you try @hsubox76's suggestion?

**Comment by emmanuel-2u** [comment]
Just stumbled upon this thread and I happen to have the same code as @hsubox76 suggested. I imported `initializeAuth` from `firebase/auth/react-native` but the problem persists. During fast refresh, `auth/already-initialized` error gets thrown.

**Comment by hsubox76** [comment]
Can you show all of your relevant code, including all Firebase imports and where you call initializeApp and initializeAuth?

**Comment by emmanuel-2u** [comment]
**Code**

``` javascript
import { initializeApp } from 'firebase/app';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {
    initializeAuth,
    getReactNativePersistence
} from 'firebase/auth/react-native';

const firebaseConfig = {
    apiKey: "apiKey",
    authDomain: "authDomain",
    projectId: "projectId",
    storageBucket: "storageBucket",
    messagingSenderId: "messagingSenderId",
    appId: "appId"
};

const firebaseApp = initializeApp(firebaseConfig);

const firebaseAuth = initializeAuth(firebaseApp, {
    persistence: getReactNativePersistence(AsyncStorage)
});
```

**Comment by Slapbox** [comment]
I haven't gotten a chance to try @hsubox76's suggestion because I'm pretty sure we can't use it. Our code is cross-platform, so I don't think we can easily implement the suggestion, but maybe via `babel-module-resolver-plugin`. I'll try to find time to test it in the near future.

**Comment by emmanuel-2u** [comment]
I noticed the checks and I've tried it before commenting but it didn't work. I added the `console.log` and noticed it's getting called again.

However, this worked for me
```javascript
import { getAuth, initializeAuth, getReactNativePersistence} from "firebase/auth/react-native";
import AsyncStorage from "@react-native-async-storage/async-storage";

if (getApps().length === 0) {
app = initializeApp(firebaseConfig);
auth = initializeAuth(app, {
persistence: getReactNativePersistence(AsyncStorage),
});

} else {
app = getApp();
auth = getAuth(app);
}
```
[https://github.com/firebase/firebase-js-sdk/issues/6050#issuecomment-1119131377](Scooter1337)

**Comment by hsubox76** [comment]
Sorry to hijack this thread but I'm trying to make sure we correct our documentation to be clear about how to use `firebase/auth/react-native` to prevent people from going through these headaches unnecessarily, and I don't seem to be able to find the documentation for it. Could anyone using that package tell me where you found out about `firebase/auth/react-native`? Was it just from Github issues or looking at the Firebase code itself, or is there a docs page somewhere I missed? If there aren't any docs I'll try to add some.

**Comment by Slapbox** [comment]
@hsubox76 thanks for your continued attention on this issue.

> Could anyone using that package tell me where you found out about firebase/auth/react-native?

I first became aware of it from comments on issue, which do link to some mentions in the API reference: #6050 - [Specifically here](https://github.com/firebase/firebase-js-sdk/issues/6050#issuecomment-1060918388) and in comments below.

**Comment by tranlammedia** [comment]
this í my sovle. use {initializeAuth} from 'firebase/auth/react-native' replace for {getAuth} from "firebase/auth";

`import AsyncStorage from '@react-native-async-storage/async-storage';
import { initializeApp } from 'firebase/app';
import {
  initializeAuth,
  getReactNativePersistence
} from 'firebase/auth/react-native';

// add firebase config here

// initialize firebase app
const app = initializeApp(firebaseConfig);

// initialize auth
const auth = initializeAuth(app, {
  persistence: getReactNativePersistence(AsyncStorage)
});

export { auth };`

**Comment by martinivanalfonso** [comment]
This seems to be fixed on version 10.0.0 

thank you @hsubox76

Sidenote for anyone following this thread, remove this fix if implemented as it seems to break fast refreshing otherwise

`const auth = initializeAuth(app, {
  persistence: getReactNativePersistence(AsyncStorage)
});`

**Comment by hsubox76** [comment]
Thanks for the update. I'll close this issue for now and add a link to the release notes for 10.0.0 to make it clear that

1) We got rid of the special explicit path for the react-native bundle and added a `react-native` field to the package.json so that the react-native should be automatically chosen by the Metro bundler
2) The AsyncStorage community package is now a dependency of firebase/auth and should now be the default if you call getAuth() in react native

https://firebase.google.com/support/release-notes/js#version_1000_-_july_6_2023
